### PR TITLE
feat: handle multiple http servers with one graceful service

### PR DESCRIPTION
**Pull request by sc-zenokerr**
Prerequisite to fix sand issue: https://github.com/Scalingo/sand/issues/271 (since graceful v1.1.2)
Related to PR #909
Related to Story STORY-828 

- [ ] Add a changelog entry in `CHANGELOG.md`

### CODE REVIEW

**Review by leo-scalingo — COMMENTED** [review]
A small comment:

Do we really have to define the number of servers in the constructor as option. Can't it be counted when `s.ListenAndServe` is called?

**Comment by sc-zenokerr** [comment]
We need to know "when" to call the tableflip upgrader "Ready()" method.

I saw three options:

1. Add a method to the API to tell graceful that all listeners have been added
2. Add a config option to let graceful know in advance how many httpServers to expect
3. Add a short timeout so that all httpServers must be registered before the timeout

I chose option 2 because:

* minimal change to the API
* it is quite clear

I am happy to change it if you have any better ideas / preferences.

In the meantime, there is a "race" in this PR, so I am going to fix that, and send an update shortly.

**Comment by leo-scalingo** [comment]
> We need to know "when" to call the tableflip upgrader "Ready()" method.
> 
> I saw three options:
> 
>     1. Add a method to the API to tell graceful that all listeners have been added
> 
>     2. Add a config option to let graceful know in advance how many httpServers to expect
> 
>     3. Add a short timeout so that all httpServers must be registered before the timeout
> 
> 
> I chose option 2 because:
> 
>     * minimal change to the API
> 
>     * it is quite clear
> 
> 
> I am happy to change it if you have any better ideas / preferences.
> 
> In the meantime, there is a "race" in this PR, so I am going to fix that, and send an update shortly.

Really good explanation, LGTM

**Review by leo-scalingo — APPROVED** [review]
Validating the chosen methodology, letting Pierre for the code analysis :)

**Comment by sc-zenokerr** [comment]
@curzolapierre this is ready for code review now.

File: graceful/service.go [hunk 1]
```diff
@@ -175,15 +239,47 @@ func (s *Service) DecConnCount(ctx context.Context) {
...
+
+	for i, httpServer := range s.httpServers {
+		wg.Add(1)
+		go func(i int, httpServer *http.Server) {
+			defer wg.Done()
+			log := logger.Get(ctx)
+			if len(s.httpServers) > 1 {
+				log = log.WithField("index", i)
+			}
+			log.Info("shutting down http server")
```
Comment (curzolapierre) [inline graceful/service.go:254]:
```suggestion
			log.Info("Shutting down http server")
```

File: graceful/service.go [hunk 2]
```diff
@@ -175,15 +239,47 @@ func (s *Service) DecConnCount(ctx context.Context) {
...
+			err := httpServer.Shutdown(ctx)
+			if err != nil {
+				log.WithError(err).Error("fail to shutdown http server")
+				errChan <- errors.Wrapf(ctx, err, "fail to shutdown http server %d", i)
+			} else {
+				log.Info("http server is stopped")
```
Comment (curzolapierre) [inline graceful/service.go:260]:
```suggestion
				log.Info("Http server is stopped")
```

File: graceful/service.go [hunk 3]
```diff
@@ -175,15 +239,47 @@ func (s *Service) DecConnCount(ctx context.Context) {
...
+			shutdownErr = errors.Wrap(ctx, shutdownErr, err.Error())
+		}
+	}
+
+	if shutdownErr != nil {
+		return shutdownErr
 	}
-	log.Info("http server is stopped")
 
 	log.Info("wait hijacked connections")
```
Comment (curzolapierre) [inline graceful/service.go:281]:
```suggestion
	log.Info("Wait hijacked connections")
```

File: graceful/service.go [hunk 4]
```diff
@@ -175,15 +239,47 @@ func (s *Service) DecConnCount(ctx context.Context) {
...
+		go func(i int, httpServer *http.Server) {
+			defer wg.Done()
+			log := logger.Get(ctx)
+			if len(s.httpServers) > 1 {
+				log = log.WithField("index", i)
+			}
+			log.Info("shutting down http server")
+			err := httpServer.Shutdown(ctx)
+			if err != nil {
+				log.WithError(err).Error("fail to shutdown http server")
```
Comment (curzolapierre) [inline graceful/service.go:257]:
```suggestion
				log.WithError(err).Error("Fail to shutdown http server")
```

File: graceful/service.go [hunk 5]
```diff
@@ -175,15 +239,47 @@ func (s *Service) DecConnCount(ctx context.Context) {
...
+				errChan <- errors.Wrapf(ctx, err, "fail to shutdown http server %d", i)
```
Comment (curzolapierre) [inline graceful/service.go:258]:
```suggestion
				errChan <- errors.Wrapf(ctx, err, "shutdown http server %d", i)
```

File: graceful/service.go [hunk 6]
```diff
@@ -117,36 +135,82 @@ func (s *Service) listenAndServe(ctx context.Context, _ string, addr string, ser
...
+
+	return nil
+}
+
+// prepare is called before the first server is started.
+func (s *Service) prepare(ctx context.Context) error {
+	if s.pidFile != "" {
+		err := os.Remove(s.pidFile)
+		if err != nil && !os.IsNotExist(err) {
+			return errors.Wrap(ctx, err, "fail to remove PID file")
```
Comment (curzolapierre) [inline graceful/service.go:161]: We don't begin error message by "fail to" any more
```suggestion
			return errors.Wrap(ctx, err, "remove PID file")
```

**Review by curzolapierre — APPROVED** [review]
Just some comments about style of logs and errors.

Otherwise LGTM

**Comment by sc-zenokerr** [comment]
@EtienneM Before I merge this, I would appreciate a discussion about it if you have time. Thanks in advance

**Pull request merged**